pkg/agent/manager/cache/cache.go

```diff
@@ -2,357 +2,582 @@ package cache
+	for entry := range entries {
+		for _, s := range entry.regEntry.Selectors {
+			if !set.In(s) {
+				delete(entries, entry)
+			}
+		}
 	}
-	return
+	return entries
 }
 
-func (c *cacheImpl) FetchWorkloadUpdate(selectors Selectors) *WorkloadUpdate {
-	entries := c.Entries()
-	bundles := c.Bundles()
+// getSelectorIndex gets a selector index entry for the selector. If one
+// doesn't exist, it is created.
+func (c *Cache) getSelectorIndex(s selector) *selectorIndex {
+	index, ok := c.selectors[s]
+	if !ok {
+		index = newSelectorIndex()
+		c.selectors[s] = index
+	}
+	return index
+}
 
-	return c.makeWorkloadUpdate(selectorsEntries(selectors, entries), bundles)
+type cacheEntry struct {
+	regEntry *common.RegistrationEntry
+	svid     *X509SVID
+	subs     map[*subscriber]struct{}
 }
 
-func (c *cacheImpl) GetJWTSVID(spiffeID string, audience []string) (*client.JWTSVID, bool) {
-	key := keyFromJWTSpiffeIDAndAudience(spiffeID, audience)
-	c.m.Lock()
-	defer c.m.Unlock()
-	svid, ok := c.jwtSVIDS[key]
-	return svid, ok
+func newCacheEntry() *cacheEntry {
+	return &cacheEntry{
+		subs: make(map[*subscriber]struct{}),
+	}
 }
 
-func (c *cacheImpl) SetJWTSVID(spiffeID string, audience []string, svid *client.JWTSVID) {
-	key := keyFromJWTSpiffeIDAndAudience(spiffeID, audience)
+type selectorIndex struct {
+	// subscriptions related to this selector
+	subs map[*subscriber]struct{}
 
-	c.m.Lock()
-	defer c.m.Unlock()
-	c.jwtSVIDS[key] = svid
+	// cache entries related to this selector
+	entries map[*cacheEntry]struct{}
 }
 
-func subscriberEntries(sub *subscriber, entries []*Entry) (subentries []*Entry) {
-	return selectorsEntries(sub.sel, entries)
+func (x *selectorIndex) isEmpty() bool {
+	return len(x.subs) == 0 && len(x.entries) == 0
 }
 
-func selectorsEntries(selectors Selectors, entries []*Entry) (subentries []*Entry) {
-	for _, e := range entries {
-		regEntrySelectors := selector.NewSetFromRaw(e.RegistrationEntry.Selectors)
-		if selector.NewSetFromRaw(selectors).IncludesSet(regEntrySelectors) {
-			subentries = append(subentries, e)
-		}
+func newSelectorIndex() *selectorIndex {
+	return &selectorIndex{
+		subs:    make(map[*subscriber]struct{}),
+		entries: make(map[*cacheEntry]struct{}),
 	}
-	return subentries
 }
 
-func certsEqual(a, b []*x509.Certificate) bool {
-	if len(a) != len(b) {
-		return false
-	}
-
-	for i, cert := range a {
-		if !cert.Equal(b[i]) {
-			return false
-		}
-	}
-
-	return true
+func sortEntries(entries []Entry) {
+	sort.Slice(entries, func(a, b int) bool {
+		return entries[a].RegistrationEntry.EntryId < entries[b].RegistrationEntry.EntryId
+	})
 }
 
-func keyFromJWTSpiffeIDAndAudience(spiffeID string, audience []string) string {
-	h := sha1.New()
-
-	// duplicate and sort the audience slice before sorting
-	audience = append([]string(nil), audience...)
-	sort.Strings(audience)
-
-	io.WriteString(h, spiffeID)
-	for _, a := range audience {
-		io.WriteString(h, a)
+func makeEntry(entry *cacheEntry) Entry {
```

Comment:
We should qualify all of these `entry`s e.g. sortCacheEntries, makeCacheEntry, etc. Some of this logic is really hard to follow without a distinction

Reply:
👍